### ISSUE TITLE
[need advice] no completion-at-point possible with gnus and notmuch?

### ISSUE THREAD

**Issue by poulpoulsen** [girzel/ebdb#104]
hello,
i use your package about a year and it contains all that i need.
unfortunately since some time i can't complete at point with <TAB> inside the message-buffer eg on the To: Header.
if i call (ebdb-complete-enable) a window pops up and i can select a candidate.
but i want completion-at-point.
i set ebdb-complete-mail to capf but without luck.

could you give an advice how to complete-at-point?
i use emacs 29 and your newest version on Linux Fedora 35.

inside using notmuch there is also no completion-at-point.

regards
poul

**Comment by girzel** [comment]
Hi there! If you just set `ebdb-complete-mail` to `t`, then compose a message, what is the value of `message-completion-alist`?

It should have an element in it like: `("^\\(Resent-\\)?\\(To\\|B?Cc\\|Reply-To\\|From\\|Mail-Followup-To\\|Mail-Copies-To\\):" . ebdb-complete-mail)`

Then the <TAB> key should run `message-tab`. Are any of those things not true?

**Comment by poulpoulsen** [comment]
hello there,
my ebdb-complete-mail usualy is set to capf but now i set it to t.
my message-completion-alist is 
(("^\\(Resent-\\)?\\(To\\|B?Cc\\|Reply-To\\|From\\|Mail-Followup-To\\|Mail-Copies-To\\):" . ebdb-complete-mail)

if i push tab a window opens with the address.
in the past i've got a completions-list of candidates at point.
like capf in emacs the usual way.
anything has changed...
perhaps something inside emacs?
i use emacs 29...

Regards
poul

**Comment by girzel** [comment]
When you say "a window opens with the address", do you mean it pops up an `*EBDB*` buffer? I'm still having trouble picturing what's actually happening. As far as I know, nothing has changed that would do this. You're sure <TAB> is bound to `message-tab`, not `ebdb-complete-message-tab`? Are you calling `ebdb-complete-enable` anywhere?

**Comment by poulpoulsen** [comment]
you are right. i called ebdb-complee-enable.
after commented it out,
<tab> is bound to 
TAB (translated from <tab>) runs the command message-tab (found in
button-map), which is an interactive compiled Lisp function in
‘message.el’.

but unfortunately
ebdb in message-completion-alist is not set.
message-completion-alist is a variable defined in ‘message.el’.

Its value is
(("^\\(Resent-\\)?\\(To\\|B?Cc\\|Reply-To\\|From\\|Mail-Followup-To\\|Mail-Copies-To\\):" . ebdb-complete-mail)
 ("^\\(Newsgroups\\|Followup-To\\|Posted-To\\|Gcc\\):" . message-expand-group)
 ("^\\([^ :]*-\\)?\\(To\\|B?Cc\\|From\\|Reply-to\\|Mail-Followup-To\\|Mail-Copies-To\\):" . message-expand-name))
Original value was 
(("^\\(Newsgroups\\|Followup-To\\|Posted-To\\|Gcc\\):" . message-expand-group)
 ("^\\([^ :]*-\\)?\\(To\\|B?Cc\\|From\\|Reply-to\\|Mail-Followup-To\\|Mail-Copies-To\\):" . message-expand-name))

perhaps ebdb is not loaded.
how could i loaded correctly?
should it load in gnus.el ?

Regards
poul

**Comment by girzel** [comment]
That's correct! The first line in your `message-completion-alist` references `ebdb-complete-mail`, that's how it's supposed to work. Try restarting Emacs with these settings; I would expect EBDB completion to work normally now.

**Comment by poulpoulsen** [comment]
Hello,
unfortunately it is not working as expected.
I will explain:
Cursor is in the To: line and i type in info, then tab key. 
Because i do have more then one adress that contains the string info, all candidates should displayed.
But only one candidate is displayed. In the past i've got a list with the candidates.
any idea?

**Comment by girzel** [comment]
Sorry I'm a bit behind on this... 

The first thing to try is to confirm that `(all-completions "info" ebdb-hashtable #'ebdb-completion-predicate)` returns all the strings you'd expect.

Then, if possible, it would be great to have you edebug the source of `ebdb-complete-mail`, and see what is happening. The part you're looking for is down the function a little ways:

```
  (let* ((end (point))
         (done (unless beg 'nothing))
         (orig (and beg (buffer-substring-no-properties beg end)))
         (completion-ignore-case t)
         (completion (and orig
                          (try-completion orig ebdb-hashtable
                                          #'ebdb-completion-predicate)))
         all-completions dwim-completions one-record)
```

You want to be sure that 1) the `orig` variable holds the string you're expecting (the fragment you're trying to complete), and that later on, the call to `all-completions` continues to return all the expected strings, and then this part:

```
      (let ((records (delete-dups
                      (apply #'append (mapcar (lambda (compl)
					        (gethash compl ebdb-hashtable))
					      all-completions)))))
```

Actually returns records. Sorry that's a bunch of random recommendations, but I expect that something in there will show us what's going wrong.

**Comment by poulpoulsen** [comment]
hello,
no problem for the delay :-)
but sorry, i do not fully understand, what you explained and what i have to do.
i tried that:
new message buffer.
put cursor into TO:
call M-: (all-completions "info" ebdb-hashtable #'ebdb-completion-predicate)
 ("info" "info@baumschule-lauer.de" "info@pflanzen-centrum-freienwill.de")

this is strange, because these adresses are inside the ebdb-db but why are they selected???

can i help?

Regards
Poul

**Comment by poulpoulsen** [comment]
i do not understand the paragraph about edebug???

**Comment by poulpoulsen** [comment]
hello,
hmmh it is not working as expected :-(
First 
C-m-x is not defined here. I called eval-buffer and that works.
Second
if i complete names and push TAB it do not open Completion Buffer, its a EBDB-Message Buffer.
Perhaps because i activate ebdb with
(add-hook 'gnus-startup-hook #'ebdb-complete-enable)
.
But if i comment it out, i got no completion.
ebdb function is not bound to TAB.
so anything from ebdb should be loaded (with hook?) or should i set a anything special.

here are my settings:
(require 'ebdb)
(require 'ebdb-gnus)
(require 'ebdb-message)
(require 'ebdb-complete)
(require 'ebdb-format)
(require 'ebdb-org)
(require 'ebdb-snarf)
(require 'ebdb-vcard)
(require 'ebdb-i18n)
(require 'ebdb-html)

(setq ebdb-sources (locate-user-emacs-file "ebdb"))
(setq ebdb-permanent-ignores-file (locate-user-emacs-file "ebdb-permanent-ignores"))
(setq ebdb-mua-pop-up nil)
(setq ebdb-default-window-size 0.25)
(setq ebdb-mua-default-formatter ebdb-default-multiline-formatter)
(setq ebdb-mua-auto-update-p 'existing)
(setq ebdb-gnus-auto-update-p 'existing)
(setq ebdb-mua-reader-update-p 'existing)
(setq ebdb-mua-sender-update-p 'query)
(setq ebdb-message-auto-update-p 'query)
(setq ebdb-message-try-all-headers t)
(setq ebdb-message-headers
      '((sender "From" "Resent-From" "Reply-To" "Sender")
        (recipients "Resent-To" "Resent-Cc" "Resent-CC" "To" "Cc" "CC" "Bcc" "BCC")))
(setq ebdb-message-all-addresses t)
(setq ebdb-complete-mail 'capf)
(setq ebdb-mail-avoid-redundancy nil)
(setq ebdb-completion-display-record t)
(setq ebdb-complete-mail-allow-cycling t)
(setq ebdb-record-self nil)
(setq ebdb-user-name-address-re 'self) ; match the above
(setq ebdb-save-on-exit t)
(setq ebdb-use-diary nil)

(let ((map ebdb-mode-map))
  (define-key map (kbd "D") #'ebdb-delete-field-or-record)
  (define-key map (kbd "M") #'ebdb-email) ; disables `ebdb-mail-each'
  (define-key map (kbd "m") #'ebdb-toggle-record-mark)
  (define-key map (kbd "t") #'ebdb-toggle-all-record-marks)
  (define-key map (kbd "T") #'ebdb-toggle-records-format) ; disables `ebdb-toggle-all-records-format'
  (define-key map (kbd "U") #'ebdb-unmark-all-records))
 
Hope it helps
Poul

**Comment by girzel** [comment]
I guess the first thing I would do here is remove all config that is setting variables to their default values! There's no sense in repeating those settings, and will shadow changes to the defaults in the future. For instance `ebdb-record-self` is nil by default, just leave that out unless you want to set it to the record that represents you.

Then the first thing would be to remove `(require 'ebdb-complete)`, as you've stated you don't want to use that completion interface. And yes, definitely remove `(add-hook 'gnus-startup-hook #'ebdb-complete-enable)`. Also remove the line `(setq ebdb-complete-mail 'capf)`, let's just leave that at its default to try to get basic behavior working.

Then restart Emacs.

One thing to try here is to start composing a message, go to the "To:" header, enter a string that you know should complete to multiple records, and run "M-x ebdb-complete-mail". Do you get a completion buffer? That should tell us whether the problem is inside the EBDB machinery, or outside of it.
